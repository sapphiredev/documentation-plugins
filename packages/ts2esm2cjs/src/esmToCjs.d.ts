interface RunTransformOptions {
	quote?: 'single' | 'double';
	lenDestructure?: number;
	lenModuleName?: number;
	lenIdentifier?: number;
	indent?: number;
}

export function runTransform(code: string, options: RunTransformOptions): string;
