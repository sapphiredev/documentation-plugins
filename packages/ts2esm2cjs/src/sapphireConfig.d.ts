declare module '@sapphire/prettier-config' {
	import type { Options } from 'prettier';

	const options: Options;

	export default options;
}
