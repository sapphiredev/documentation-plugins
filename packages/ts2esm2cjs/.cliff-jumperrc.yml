name: docusaurus-plugin-ts2esm2cjs
org: sapphire
install: true
packagePath: packages/ts2esm2cjs
identifierBase: false
pushTag: true
githubRelease: true
githubReleaseLatest: true
githubRepo: sapphiredev/documentation-plugins
