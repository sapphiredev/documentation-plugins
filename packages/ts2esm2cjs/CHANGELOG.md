# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [1.0.2](https://github.com/sapphiredev/documentation-plugins/compare/@sapphire/docusaurus-plugin-ts2esm2cjs@1.0.1...@sapphire/docusaurus-plugin-ts2esm2cjs@1.0.2) (2022-01-12)

### Bug Fixes

-   **ts2esm2cjs:** fixed imports that have `as` in their name ([147c85d](https://github.com/sapphiredev/documentation-plugins/commit/147c85db866ac0544b4bcc62aa5e54a9f3009903))

### Reverts

-   Revert "ci: try adding extra dev deps for ci" ([de49260](https://github.com/sapphiredev/documentation-plugins/commit/de49260d4d8c4be4bb27d53f13472946e4cd3700))

## [1.0.1](https://github.com/sapphiredev/documentation-plugins/compare/@sapphire/docusaurus-plugin-ts2esm2cjs@1.0.0...@sapphire/docusaurus-plugin-ts2esm2cjs@1.0.1) (2021-11-30)

### Bug Fixes

-   **ts2esm2cjs:** allow longer lengths for esm to cjs ([280746c](https://github.com/sapphiredev/documentation-plugins/commit/280746cbf7120c9baddf8deac436ea0a9ba1835e))

# 1.0.0 (2021-11-26)

### Bug Fixes

-   **ts2esm2cjs:** removed esm exports to ensure cjs works fine with require syntax ([73b6898](https://github.com/sapphiredev/documentation-plugins/commit/73b6898881374259f9a38b74fe741cf46a468e06))
-   **ts2esm2cjs:** removed exports field from package.json ([7dfac94](https://github.com/sapphiredev/documentation-plugins/commit/7dfac9464b28caa8e2d0dc03040a1b38f6b02d68))

### Features

-   add ts2esm2cjs package ([35ec87d](https://github.com/sapphiredev/documentation-plugins/commit/35ec87dd8743aecc57c344b1d2d4ae88038fcd7b))
-   **ts2esm2cjs:** add option to customize tsc options ([af19a1d](https://github.com/sapphiredev/documentation-plugins/commit/af19a1d8473b8a3723933f87f8a0c6f79e98ebc7))

### Reverts

-   Revert "fix(typedoc-djs-links): removed esm exports to ensure cjs works fine with require syntax" ([7b4c6ab](https://github.com/sapphiredev/documentation-plugins/commit/7b4c6ab434265903213d7d965b0e927912b1aba1))
