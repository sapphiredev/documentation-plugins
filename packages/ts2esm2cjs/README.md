<div align="center">

![Sapphire Logo](https://cdn.skyra.pw/gh-assets/sapphire-banner.png)

# @sapphire/docusaurus-plugin-ts2esm2cjs

**Docusaurus Remark plugin for converting TypeScript code to ESM and CJS code.**

[![GitHub](https://img.shields.io/github/license/sapphiredev/documentation-plugins)](https://github.com/sapphiredev/documentation-plugins/blob/main/LICENSE.md)
[![npm](https://img.shields.io/npm/v/@sapphire/docusaurus-plugin-ts2esm2cjs?color=crimson&logo=npm&style=flat-square)](https://www.npmjs.com/package/@sapphire/docusaurus-plugin-ts2esm2cjs)

</div>

## Description

This is a plugin for [Docusaurus](https://docusaurus.io) that adds support for a `ts2esmcjs` meta tag for code blocks.
It will transpile your TypeScript code into ESM and CJS code, and format it properly. It will then give you a component
using `Tabs` where users can select their preferred language. By default this will sync across all code blocks that use
this meta tag, you can disable that through the plugin options.

## Usage

In your `docusaurus.config.js`, for the plugins where you need this feature (doc, blog, pages, etc.), register it in the
`remarkPlugins` option. (See [Docs configuration][docconf] for more details on configuration format)

```js
// docusaurus.config.js
module.exports = {
  // ...
  presets: [
    [
      '@docusaurus/preset-classic',
      {
        docs: {
          remarkPlugins: [[require('@sapphire/docusaurus-plugin-ts2esm2cjs')]]
        },
        pages: {
          remarkPlugins: [require('@sapphire/docusaurus-plugin-ts2esm2cjs')]
        },
        blog: {
          // ...
        }
      }
    ]
  ]
};
```

And then use it by adding the `ts2esm2cjs` key to the code block:

**meta**: `_typescript ts2esm2cjs_`

```typescript
import { Command, Args } from '@sapphire/framework';
import type { Message } from 'discord.js';

export class EchoCommand extends Command {
  public constructor(context: Command.Context, options: Command.Options) {
    super(context, {
      ...options,
      aliases: ['parrot', 'copy'],
      description: 'Replies with the text you provide'
    });
  }

  public async messageRun(message: Message, args: Args) {
    // ...
  }
}
```

### Line Highlighting

As you may be aware, Docusaurus supports [Line Highlighting][line-highlighting] for code blocks. As line highlighting
relies on meta just like this plugin would, we would normally run into a bit of a conflict as to what to set in the meta
tag. On top of that, often you'll want to highlight different lines for JavaScript as opposed to TypeScript. To solve
these issues this plugin supports a unique syntax of setting the meta tag. You can configure the line highlighting by
adding a `|` character after the plugin meta tag. This can be done twice, wherein the first line highlight configuration
will be for the JavaScript and ESM code blocks, whereas the second will be for the TypeScript code block.

For example consider the following code block:

**meta:** `_typescript ts2esm2cjs|{1,4-10}|{1,5-11}_ `

```typescript
import { Command, Args } from '@sapphire/framework';
import type { Message } from 'discord.js';

export class EchoCommand extends Command {
  public constructor(context: Command.Context, options: Command.Options) {
    super(context, {
      ...options,
      aliases: ['parrot', 'copy'],
      description: 'Replies with the text you provide'
    });
  }

  public async messageRun(message: Message, args: Args) {
    // ...
  }
}
```

This will give you a three code blocks in a `Tabs` layout with the JavaScript and ESM code blocks having lines 1 and 4
through 10 highlighted, and the TypeScript code block having lines 1 and 5 through 11 highlighted.

### Code formatting

The output JS and ESM code is formatted using Prettier. The default configuration is:

```json
{
  "endOfLine": "lf",
  "printWidth": 120,
  "quoteProps": "as-needed",
  "semi": true,
  "singleQuote": true,
  "tabWidth": 2,
  "trailingComma": "none",
  "useTabs": false,
  "parser": "babel"
}
```

You can customize this by passing `prettierConfig` to the plugin options.

### TypeScript compiler options

The ESM code is transpiled from TypeScript using the TypeScript compiler with the following base configuration:

```json
{
  "newLine": "lf",
  "removeComments": false,
  "esModuleInterop": true,
  "pretty": true,
  "module": "ESNext",
  "moduleResolution": "Node",
  "target": "ESNext"
}
```

You can customize this by passing `typescriptCompilerOptions` to the plugin options. However, you cannot customize the
`module`, `moduleResolution`, or `target` options.

## Buy us some doughnuts

Sapphire Community is and always will be open source, even if we don't get donations. That being said, we know there are
amazing people who may still want to donate just to show their appreciation. Thank you very much in advance!

We accept donations through Open Collective, Ko-fi, Paypal, Patreon and GitHub Sponsorships. You can use the buttons
below to donate through your method of choice.

|   Donate With   |                       Address                       |
| :-------------: | :-------------------------------------------------: |
| Open Collective | [Click Here](https://sapphirejs.dev/opencollective) |
|      Ko-fi      |      [Click Here](https://sapphirejs.dev/kofi)      |
|     Patreon     |    [Click Here](https://sapphirejs.dev/patreon)     |
|     PayPal      |     [Click Here](https://sapphirejs.dev/paypal)     |

## Contributors âœ¨

Thanks goes to these wonderful people ([emoji key](https://allcontributors.org/docs/en/emoji-key)):

<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<!-- prettier-ignore-start -->
<!-- markdownlint-disable -->
<table>
  <tr>
    <td align="center"><a href="https://favware.tech/"><img src="https://avatars3.githubusercontent.com/u/4019718?v=4?s=100" width="100px;" alt=""/><br /><sub><b>Jeroen Claassens</b></sub></a><br /><a href="https://github.com/sapphiredev/documentation-plugins/commits?author=favna" title="Code">ðŸ’»</a> <a href="#infra-favna" title="Infrastructure (Hosting, Build-Tools, etc)">ðŸš‡</a> <a href="#projectManagement-favna" title="Project Management">ðŸ“†</a></td>
    <td align="center"><a href="https://github.com/apps/renovate"><img src="https://avatars.githubusercontent.com/in/2740?v=4?s=100" width="100px;" alt=""/><br /><sub><b>renovate[bot]</b></sub></a><br /><a href="#maintenance-renovate[bot]" title="Maintenance">ðŸš§</a></td>
    <td align="center"><a href="https://renovate.whitesourcesoftware.com/"><img src="https://avatars.githubusercontent.com/u/25180681?v=4?s=100" width="100px;" alt=""/><br /><sub><b>WhiteSource Renovate</b></sub></a><br /><a href="#maintenance-renovate-bot" title="Maintenance">ðŸš§</a></td>
  </tr>
</table>

<!-- markdownlint-restore -->
<!-- prettier-ignore-end -->

<!-- ALL-CONTRIBUTORS-LIST:END -->

This project follows the [all-contributors](https://github.com/all-contributors/all-contributors) specification.
Contributions of any kind welcome!

[docconf]: https://docusaurus.io/docs/api/plugins/@docusaurus/plugin-content-docs#ex-config
[line-highlighting]: https://docusaurus.io/docs/markdown-features/code-blocks#line-highlighting
