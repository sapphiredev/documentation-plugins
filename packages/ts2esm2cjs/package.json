{
	"name": "@sapphire/docusaurus-plugin-ts2esm2cjs",
	"version": "1.0.2",
	"description": "Docusaurus Remark plugin for converting TypeScript code to ESM and CJS code",
	"author": "@sapphire",
	"license": "MIT",
	"main": "dist/index.js",
	"module": "dist/index.mjs",
	"types": "dist/index.d.ts",
	"exports": {
		"import": "./dist/index.mjs",
		"require": "./dist/index.js"
	},
	"scripts": {
		"lint": "eslint src --ext ts --fix",
		"build": "tsup && tsc -b src",
		"typecheck": "tsc -p tsconfig.eslint.json",
		"prepublishOnly": "yarn build"
	},
	"dependencies": {
		"@sapphire/prettier-config": "^1.2.8",
		"prettier": "^2.5.1",
		"typescript": "^4.5.4"
	},
	"repository": {
		"type": "git",
		"url": "git+https://github.com/sapphiredev/documentation-plugins.git",
		"directory": "packages/ts2esm2cjs"
	},
	"files": [
		"dist",
		"!dist/*.tsbuildinfo"
	],
	"engines": {
		"node": ">=v16.6.0",
		"npm": ">=7.0.0"
	},
	"keywords": [
		"docusaurus",
		"docusaurus-plugin",
		"@sapphire",
		"typescript",
		"ts",
		"esm",
		"cjs",
		"discord",
		"sapphire",
		"discordjs"
	],
	"bugs": {
		"url": "https://github.com/sapphiredev/documentation-plugins/issues"
	},
	"homepage": "https://www.sapphirejs.dev",
	"publishConfig": {
		"access": "public"
	}
}
