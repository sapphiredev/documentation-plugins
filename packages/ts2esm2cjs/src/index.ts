import sapphirePrettierConfig from '@sapphire/prettier-config';
import { runTransform } from './esmToCjs';
import prettier, { Options } from 'prettier';
import ts, { CompilerOptions } from 'typescript';

export const documentationPrettierConfig: Options = {
	...sapphirePrettierConfig,
	tabWidth: 2,
	useTabs: false,
	printWidth: 120,
	parser: 'babel'
};

export const makeTsCompilerOptions = (overrideOptions?: CompilerOptions): CompilerOptions => ({
	newLine: ts.NewLineKind.LineFeed,
	removeComments: false,
	esModuleInterop: true,
	pretty: true,
	...overrideOptions,
	module: ts.ModuleKind.ESNext,
	moduleResolution: ts.ModuleResolutionKind.NodeJs,
	target: ts.ScriptTarget.ESNext
});

/**
 * Transpiles input TypeScript code to ESM code.
 * @param code The code to transpile
 * @returns Input code transpiled to ESM
 */
export const tsToEsm = (code: string, options: Pick<PluginOptions, 'typescriptCompilerOptions'>): ts.TranspileOutput =>
	ts.transpileModule(code, { reportDiagnostics: false, compilerOptions: makeTsCompilerOptions(options.typescriptCompilerOptions) });

/**
 * Transforms input ESM code to CJS code.
 * @param code The code to transform
 * @returns Input code transformed to CommonJS
 */
export const esmToCjs = (code: string): string =>
	runTransform(code, { quote: 'single', lenDestructure: 512, lenModuleName: 512, lenIdentifier: 512 });

/**
 * Escaped new lines in code with block comments so they can be restored by {@link restoreNewLines}
 * @param code The code to escape new lines in
 * @returns The same code but with new lines escaped using block comments
 */
export const escapeNewLines = (code: string) => code.replace(/\n\n/g, '\n/* :newline: */');

/**
 * Reverses {@link escapeNewLines} and restores new lines
 * @param code The code with escaped new lines
 * @returns The same code with new lines restored
 */
export const restoreNewLines = (code: string): string => code.replace(/\/\* :newline: \*\//g, '\n');

/**
 * Formats the code using Prettier
 * @param code The code to prettier format
 * @param prettierConfig Additional prettier options to use for formatting
 * @returns Prettier formatted code
 */
export const prettierFormatCode = (code: string, prettierConfig?: Options) =>
	prettier.format(code, { ...documentationPrettierConfig, ...prettierConfig }).slice(0, -1);

/**
 * Transforms a Docusaurus node from TypeScript to ESM and CJS
 * @param node The Docusaurus node to transform
 * @param isSync Whether the transform should synchronize between all entries of this type
 * @returns The transformed node in the form of Tabs.
 */
const transformNode = (node: any, options: PluginOptions) => {
	const groupIdProp = options.sync ? 'groupId="ts2esm2cjs" ' : '';

	const tsCode = escapeNewLines(node.value);
	const esmCode = tsToEsm(tsCode, { typescriptCompilerOptions: options.typescriptCompilerOptions }).outputText;
	const cjsCode = esmToCjs(esmCode);

	const [, jsHighlight, tsHighlight] = node.meta.split('|');

	return [
		{
			type: 'jsx',
			value:
				`<Tabs defaultValue="typescript" ${groupIdProp}` +
				`values={[
    { label: 'JavaScript', value: 'javascript', },
    { label: 'ESM', value: 'esm', },
    { label: 'TypeScript', value: 'typescript', },
  ]}
>
<TabItem value="javascript">`
		},
		{
			type: node.type,
			lang: node.lang,
			meta: jsHighlight,
			value: prettierFormatCode(restoreNewLines(cjsCode), options.prettierOptions)
		},
		{
			type: 'jsx',
			value: '</TabItem>\n<TabItem value="esm">'
		},
		{
			type: node.type,
			lang: node.lang,
			meta: jsHighlight,
			value: prettierFormatCode(restoreNewLines(esmCode), options.prettierOptions)
		},
		{
			type: 'jsx',
			value: '</TabItem>\n<TabItem value="typescript">'
		},
		{
			type: node.type,
			lang: node.lang,
			meta: tsHighlight,
			value: node.value
		},
		{
			type: 'jsx',
			value: '</TabItem>\n</Tabs>'
		}
	];
};

const matchNode = (node: any) => node.type === 'code' && typeof node.meta === 'string' && node.meta.startsWith('ts2esm2cjs');
const nodeForImport = {
	type: 'import',
	value: "import Tabs from '@theme/Tabs';\nimport TabItem from '@theme/TabItem';"
};

export interface PluginOptions {
	sync?: boolean;
	prettierOptions?: Options;
	typescriptCompilerOptions?: CompilerOptions;
}

export function ts2esm2cjs(
	{ sync = true, prettierOptions = {}, typescriptCompilerOptions = {} }: PluginOptions = {
		sync: true,
		prettierOptions: {},
		typescriptCompilerOptions: {}
	}
) {
	let transformed = false;
	let alreadyImported = false;

	const transformer = (node: any) => {
		if (node.type === 'import' && node.value.includes('@theme/Tabs')) {
			alreadyImported = true;
		}

		if (matchNode(node)) {
			transformed = true;
			return transformNode(node, { sync, prettierOptions, typescriptCompilerOptions });
		}

		if (Array.isArray(node.children)) {
			let index = 0;
			while (index < node.children.length) {
				const result = transformer(node.children[index]);
				if (result) {
					node.children.splice(index, 1, ...result);
					index += result.length;
				} else {
					index += 1;
				}
			}
		}

		if (node.type === 'root' && transformed && !alreadyImported) {
			node.children.unshift(nodeForImport);
		}

		return null;
	};

	return transformer;
}
