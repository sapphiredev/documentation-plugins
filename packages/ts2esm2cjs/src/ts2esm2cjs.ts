import sapphirePrettierConfig from '@sapphire/prettier-config';
import { runTransform } from 'esm-to-cjs';
import prettier, { Options } from 'prettier';
import ts, { CompilerOptions } from 'typescript';
import type { PluginOptions } from './types';

const makeTsCompilerOptions = (overrideOptions?: CompilerOptions): CompilerOptions => ({
	newLine: ts.NewLineKind.LineFeed,
	removeComments: false,
	esModuleInterop: true,
	pretty: true,
	...overrideOptions,
	module: ts.ModuleKind.ESNext,
	moduleResolution: ts.ModuleResolutionKind.NodeJs,
	target: ts.ScriptTarget.ESNext
});

const documentationPrettierConfig: Options = {
	...sapphirePrettierConfig,
	tabWidth: 2,
	useTabs: false,
	printWidth: 120,
	parser: 'babel'
};

/**
 * Transforms input ESM code to CJS code.
 * @param code The code to transform
 * @returns Input code transformed to CommonJS
 */
const esmToCjs = (code: string): string => runTransform(code, { quote: 'single', lenDestructure: 128, lenModuleName: 128, lenIdentifier: 128 });

/**
 * Escaped new lines in code with block comments so they can be restored by {@link restoreNewLines}
 * @param code The code to escape new lines in
 * @returns The same code but with new lines escaped using block comments
 */
const escapeNewLines = (code: string) => code.replace(/\n\n/g, '\n/* :newline: */');

/**
 * Reverses {@link escapeNewLines} and restores new lines
 * @param code The code with escaped new lines
 * @returns The same code with new lines restored
 */
const restoreNewLines = (code: string): string => code.replace(/\/\* :newline: \*\//g, '\n');

/**
 * Formats the code using Prettier
 * @param code The code to prettier format
 * @param prettierConfig Additional prettier options to use for formatting
 * @returns Prettier formatted code
 */
const prettierFormatCode = (code: string, prettierConfig?: Options) =>
	prettier.format(code, { ...documentationPrettierConfig, ...prettierConfig }).slice(0, -1);

/**
 * Transpiles input TypeScript code to ESM code.
 * @param code The code to transpile
 * @returns Input code transpiled to ESM
 */
const tsToEsm = (code: string, options: Pick<PluginOptions, 'typescriptCompilerOptions'>): ts.TranspileOutput =>
	ts.transpileModule(code, { reportDiagnostics: false, compilerOptions: makeTsCompilerOptions(options.typescriptCompilerOptions) });

/**
 * Transforms input TypeScript code to ESM code.
 * @param input The TypeScript code to transform
 * @param options The plugin options
 * @returns The transformed code (ESM)
 */
export function ts2esm(input: string, options: PluginOptions) {
	const tsCode = escapeNewLines(input);
	const esmCode = tsToEsm(tsCode, { typescriptCompilerOptions: options.typescriptCompilerOptions }).outputText;

	return prettierFormatCode(restoreNewLines(esmCode), options.prettierOptions);
}

export function esm2cjs(input: string, options: PluginOptions) {
	const cjsCode = esmToCjs(input);

	return prettierFormatCode(restoreNewLines(cjsCode), options.prettierOptions);
}
