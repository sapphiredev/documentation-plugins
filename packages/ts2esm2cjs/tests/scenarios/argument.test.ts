import { readFile } from 'node:fs/promises';
import { esm2cjs, ts2esm } from '../../src/ts2esm2cjs';
import { resolve } from 'node:path';

describe('argument', async () => {
	const inputCode = await readFile(resolve(__dirname, '../fixtures/argument.ts'), { encoding: 'utf-8' });

	test('GIVEN argument fixture THEN outputs valid ESM and CJS code', () => {
		const esmCode = ts2esm(inputCode, {});
		const cjsCode = esm2cjs(esmCode, {});

		expect(esmCode).toMatchSnapshot();
		expect(cjsCode).toMatchSnapshot();
	});
});
