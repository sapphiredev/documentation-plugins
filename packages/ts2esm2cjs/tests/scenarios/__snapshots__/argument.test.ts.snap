// Vitest Snapshot v1

exports[`argument > GIVEN argument fixture THEN outputs valid ESM and CJS code 1`] = `
"// TypeScript:
import { Argument } from '@sapphire/framework';
import { URL } from 'node:url';

export class CoreArgument extends Argument {
  constructor(context) {
    super(context, { name: 'hyperlink', aliases: ['url'] });
  }

  run(argument) {
    try {
      return this.ok(new URL(argument));
    } catch {
      return this.error(argument, 'ArgumentHyperlinkInvalidURL', 'The argument did not resolve to a valid URL.');
    }
  }
}"
`;

exports[`argument > GIVEN argument fixture THEN outputs valid ESM and CJS code 2`] = `
"// TypeScript:
const { Argument } = require('@sapphire/framework');
const { URL } = require('node:url');

class CoreArgument extends Argument {
  constructor(context) {
    super(context, { name: 'hyperlink', aliases: ['url'] });
  }

  run(argument) {
    try {
      return this.ok(new URL(argument));
    } catch {
      return this.error(argument, 'ArgumentHyperlinkInvalidURL', 'The argument did not resolve to a valid URL.');
    }
  }
}
module.exports = {
  CoreArgument
};"
`;
